doc/mod-host.1
.TH mod\-host
.SH About
.PP
mod\-host is a LV2 host for JACK, controllable via socket or command line
.PP
Currently the host supports the following LV2 features:
.RS
.IP \(bu 2
lv2core
.IP \(bu 2
atom
.IP \(bu 2
event
.IP \(bu 2
buf\-size
.IP \(bu 2
midi
.IP \(bu 2
options
.IP \(bu 2
uri\-map
.IP \(bu 2
urid
.IP \(bu 2
worker
.IP \(bu 2
presets
.RE
.PP
mod\-host is part of the MOD project
\[la]http://moddevices.com\[ra]\&.
.SH Building
.PP
mod\-host uses a simple Makefile to build the source code.
The steps to build and install are:
.PP
.RS
.nf
make
make install
.fi
.RE
.PP
You can change the base installation path passing PREFIX as argument of make.
.PP
Dependencies:
.PP
.RS
.nf
libjack\-jackd2  >= 1.9.8
liblilv         >= 0.14.2
libreadline     >= 6.2
lilv\-utils      (optional)
.fi
.RE
.SH Running
.PP
mod\-host does not startup JACK automatically, so you need to start it before
running mod\-host.
.PP
If you run mod\-host without options the process will be forked and it can only
be controlled through the socket.
The default socket port is 5555, this can be changed by passing the option
\-p (or \-\-socket\-port) to mod\-host.
.PP
The other way to control mod\-host is the interactive mode, in this case the
commands must be provided on the shell prompt.
The interactive mode has autocomplete, therefore, you can always type \fB\fC[TAB]\fR
twice any time you want a hint.
.PP
Note: When you are in the interactive mode, socket communication won't work.
.SH Options
.PP
.RS
.nf
\-v, \-\-verbose
    verbose messages
\-p, \-\-socket\-port=<port>
    socket port definition
\-i, \-\-interactive
    interactive shell mode
\-V, \-\-version
    print program version and exit
\-h, \-\-help
    print help and exit
.fi
.RE
.SH Commands (or Protocol)
.PP
The commands supported by mod\-host are:
.PP
.RS
.nf
add <lv2_uri> <instance_number>
    * add a LV2 plugin encapsulated as a jack client
    e.g.: add http://lv2plug.in/plugins/eg\-amp 0
    instance_number must be any value between 0 ~ 9999, inclusively
remove <instance_number>
    * remove a LV2 plugin instance (and also the jack client)
    e.g.: remove 0
connect <origin_port> <destination_port>
    * connect two effect audio ports
    e.g.: connect system:capture_1 effect_0:in
disconnect <origin_port> <destination_port>
    * disconnect two effect audio ports
    e.g.: disconnect system:capture_1 effect_0:in
preset_load <instance_number> <preset_uri>
    * load a preset state to given effect instance
    e.g.: preset_load 0 "http://drobilla.net/plugins/mda/presets#JX10\-moogcury\-lite"
preset_save <instance_number> <preset_name> <dir> <file_name>
    * save a preset state from given effect instance
    e.g.: preset_save 0 "My Preset" /home/user/.lv2/my\-presets.lv2 mypreset.ttl
preset_show <instance_number> <preset_uri>
    * show the preset information of requested instance / URI
    e.g.: preset_show 0 http://drobilla.net/plugins/mda/presets#EPiano\-bright
param_list <instance_number>
    * list the available controls for an instance
    e.g.: param_list 0
param_set <instance_number> <param_symbol> <param_value>
    * set a value to given control
    e.g.: param_set 0 gain 2.50
param_get <instance_number> <param_symbol>
    * get the value of the request control
    e.g.: param_get 0 gain
param_monitor <instance_number> <param_symbol> <cond_op> <value>
    * do monitoring a effect instance control port according given condition
    e.g: param_monitor 0 gain > 2.50
monitor <addr> <port> <status>
    * open a socket port to monitoring parameters
    e.g: monitor localhost 12345 1
    if status = 1 start monitoring
    if status = 0 stop monitoring
bypass <instance_number> <bypass_value>
    * toggle effect processing
    e.g.: bypass 0 1
    if bypass_value = 1 bypass effect
    if bypass_value = 0 process effect
load <file_name>
    * load a history command file
    * dummy way to save/load workspace state
    e.g.: load my_setup
save <file_name>
    * saves the history of typed commands
    * dummy way to save/load workspace state
    e.g.: save my_setup
help
    * show a help message
quit
    bye!
.fi
.RE
.PP
For each effect added one client on JACK will be created. The names of clients
follow the standard: effect\fI<instance\fPnumber>
.PP
If a valid command is executed a response is given as following:
.PP
.RS
.nf
resp <status> [value]
.fi
.RE
.PP
If status is a negative number an error has occurred. The table below shows the number of each error.
.TS
allbox;
cb cb
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
l l
.
status	error
\-1	ERR_INSTANCE_INVALID
\-2	ERR_INSTANCE_ALREADY_EXISTS
\-3	ERR_INSTANCE_NON_EXISTS
\-101	ERR_LV2_INVALID_URI
\-102	ERR_LV2_INSTANTIATION
\-103	ERR_LV2_INVALID_PARAM_SYMBOL
\-104	ERR_LV2_INVALID_PRESET_URI
\-105	ERR_LV2_CANT_LOAD_STATE
\-201	ERR_JACK_CLIENT_CREATION
\-202	ERR_JACK_CLIENT_ACTIVATION
\-203	ERR_JACK_CLIENT_DEACTIVATION
\-204	ERR_JACK_PORT_REGISTER
\-205	ERR_JACK_PORT_CONNECTION
\-206	ERR_JACK_PORT_DISCONNECTION
\-301	ERR_MEMORY_ALLOCATION
.TE
.PP
A status zero or positive means that the command was executed successfully.
In case of the add command, the status returned is the instance number.
The value field currently only exists for the param_get command.


CHANGELOG
0.10.5
======

  - new: lv2 presets support

0.10.4
======
  - new: cpu_load information

0.10.3
======
  - fix: sampleRate property implementation

0.10.2
======
  - fix: calculates bounds when the port has the sampleRate property

0.10.1
======

  - fix: checking last modified parameter

0.10
====

  - new: man pages files
  - new: control set parameter optimization
  - new: documentation and autocomplete to monitor command
  - fix: command print in verbose mode

0.9
===

  - All current code was named v0.9


src/mod-host.c
/*
 * This file is part of mod-host.
 *
 * mod-host is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mod-host is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mod-host.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

/*
************************************************************************************************************************
*
************************************************************************************************************************
*/


/*
************************************************************************************************************************
*           INCLUDE FILES
************************************************************************************************************************
*/

#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <jack/jack.h>
#include <readline/readline.h>
#include <readline/history.h>
#include <signal.h>

#ifdef HAVE_FFTW335
#include <fftw3.h>
#endif

#ifdef HAVE_NE10
#include <NE10.h>
#endif

#ifdef HAVE_NEW_LILV
#include <lilv/lilv.h>
#else
#define lilv_free(x) free(x)
#endif

#include "mod-host.h"
#include "effects.h"
#include "socket.h"
#include "protocol.h"
#include "completer.h"
#include "monitor.h"
#include "info.h"


/*
************************************************************************************************************************
*           LOCAL DEFINES
************************************************************************************************************************
*/

#define PID_FILE        "/tmp/mod-host.pid"


/*
************************************************************************************************************************
*           LOCAL CONSTANTS
************************************************************************************************************************
*/

/* The text of help message is extracted from README file */
extern const char help_msg[];
/* The version is extracted from git history */
extern const char version[];
/* Wherever we should be running */
static volatile int running;
/* Thread that calls socket_run() for the JACK internal client */
static pthread_t intclient_socket_thread;

/*
************************************************************************************************************************
*           LOCAL DATA TYPES
************************************************************************************************************************
*/


/*
************************************************************************************************************************
*           LOCAL MACROS
************************************************************************************************************************
*/


/*
************************************************************************************************************************
*           LOCAL GLOBAL VARIABLES
************************************************************************************************************************
*/


/*
************************************************************************************************************************
*           LOCAL FUNCTION PROTOTYPES
************************************************************************************************************************
*/


/*
************************************************************************************************************************
*           LOCAL CONFIGURATION ERRORS
************************************************************************************************************************
*/


/*
************************************************************************************************************************
*           LOCAL FUNCTIONS
************************************************************************************************************************
*/

static void effects_add_cb(proto_t *proto)
{
    int resp;
    resp = effects_add(proto->list[1], atoi(proto->list[2]));

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void effects_remove_cb(proto_t *proto)
{
    int resp;
    resp = effects_remove(atoi(proto->list[1]));

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void effects_preset_save_cb(proto_t *proto)
{
    int resp;
    resp = effects_preset_save(atoi(proto->list[1]), proto->list[3], proto->list[4], proto->list[2]);

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void effects_preset_load_cb(proto_t *proto)
{
    int resp;
    resp = effects_preset_load(atoi(proto->list[1]), proto->list[2]);

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void effects_preset_show_cb(proto_t *proto)
{
    char *state_str = NULL;
    if (effects_preset_show(proto->list[1], &state_str) == SUCCESS)
    {
        if (state_str)
        {
            protocol_response(state_str, proto);
            lilv_free(state_str);
            return;
        }
    }
    protocol_response("", proto);
}

static void effects_connect_cb(proto_t *proto)
{
    int resp;
    resp = effects_connect(proto->list[1], proto->list[2]);

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void effects_disconnect_cb(proto_t *proto)
{
    int resp;
    resp = effects_disconnect(proto->list[1], proto->list[2]);

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void effects_bypass_cb(proto_t *proto)
{
    int resp;
    resp = effects_bypass(atoi(proto->list[1]), atoi(proto->list[2]));

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void effects_set_param_cb(proto_t *proto)
{
    int resp;
    resp = effects_set_parameter(atoi(proto->list[1]), proto->list[2], atof(proto->list[3]));
    if (resp != SUCCESS)
    {
        resp = effects_set_property(atoi(proto->list[1]), proto->list[2], proto->list[3]);
    }

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void effects_get_param_cb(proto_t *proto)
{
    int resp;
    float value;
    resp = effects_get_parameter(atoi(proto->list[1]), proto->list[2], &value);

    char buffer[128];
    if (resp >= 0)
        sprintf(buffer, "resp %i %.04f", resp, value);
    else
        sprintf(buffer, "resp %i", resp);

    protocol_response(buffer, proto);
}

static void effects_monitor_param_cb(proto_t *proto)
{
    int resp;
    if (monitor_status())
        resp = effects_monitor_parameter(atoi(proto->list[1]), proto->list[2],
                                         proto->list[3], atof(proto->list[4]));
    else
        resp = -1;

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void monitor_addr_set_cb(proto_t *proto)
{
    int resp;
    if (atoi(proto->list[3]) == 1)
        resp = monitor_start(proto->list[1], atoi(proto->list[2]));
    else
        resp = monitor_stop();

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void monitor_output_cb(proto_t *proto)
{
    int resp;
    resp = !effects_monitor_output_parameter(atoi(proto->list[1]), proto->list[2]);

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void midi_learn_cb(proto_t *proto)
{
    int resp;
    resp = !effects_midi_learn(atoi(proto->list[1]), proto->list[2]);

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void midi_map_cb(proto_t *proto)
{
    int resp;
    resp = !effects_midi_map(atoi(proto->list[1]), proto->list[2], atoi(proto->list[3]), atoi(proto->list[4]));

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void midi_unmap_cb(proto_t *proto)
{
    int resp;
    resp = !effects_midi_unmap(atoi(proto->list[1]), proto->list[2]);

    char buffer[128];
    sprintf(buffer, "resp %i", resp);
    protocol_response(buffer, proto);
}

static void midi_program_listen_cb(proto_t *proto)
{
    effects_midi_program_listen(atoi(proto->list[1]), atoi(proto->list[2]));

    char buffer[128];
    sprintf(buffer, "resp 0");
    protocol_response(buffer, proto);
}

static void cpu_load_cb(proto_t *proto)
{
    float value = effects_jack_cpu_load();
    char buffer[128];
    sprintf(buffer, "resp 0 %.04f", value);

    protocol_response(buffer, proto);
}

static void load_cb(proto_t *proto)
{
    FILE *fp;

    fp = fopen(proto->list[1], "r");
    if (fp)
    {
        char line[1024];
        msg_t msg;
        msg.sender_id = STDOUT_FILENO;

        while (fgets(line, sizeof(line), fp))
        {
            printf("%s", line);

            /* removes the \n at end of line */
            line[strlen(line)-1] = 0;

            /* fills the message struct and parse it */
            msg.data = line;
            msg.data_size = strlen(line);
            protocol_parse(&msg);

            printf("\n");
        }

        fclose(fp);
    }
    else
    {
        protocol_response("error: can't open the file", proto);
    }
}

static void save_cb(proto_t *proto)
{
    HIST_ENTRY *entry;

    if (history_length > 0)
    {
        /* removes the save command from history */
        entry = remove_history(history_length-1);
        free_history_entry(entry);

        /* saves the history in the file */
        write_history(proto->list[1]);
        protocol_response("resp 0", proto);
    }
}

static void bundle_add(proto_t *proto)
{
    effects_bundle_add(proto->list[1]);
    protocol_response("resp 0", proto);
}

static void bundle_remove(proto_t *proto)
{
    effects_bundle_remove(proto->list[1]);
    protocol_response("resp 0", proto);
}

static void data_handled(proto_t *proto)
{
    effects_data_handled();
    protocol_response("resp 0", proto);
}

static void help_cb(proto_t *proto)
{
    proto->response = 0;

    size_t i, len = strlen(help_msg);
    for (i = 0; i < len; i++)
        printf("%c", help_msg[i]);
    fflush(stdout);
}

static void quit_cb(proto_t *proto)
{
    protocol_response("resp 0", proto);

    protocol_remove_commands();
    socket_finish();
    effects_finish(1);
    exit(EXIT_SUCCESS);
}

static void interactive_mode(void)
{
    msg_t msg;
    char *input;

    completer_init();

    msg.sender_id = STDOUT_FILENO;

    while (1)
    {
        input = readline("mod-host> ");
        rl_bind_key('\t', rl_complete);

        if (input)
        {
            if (*input)
            {
                /* adds the line on history */
                add_history(input);

                /* fills the message struct and parse it */
                msg.data = input;
                msg.data_size = strlen(input);
                protocol_parse(&msg);
                printf("\n");

                free(input);
                input = NULL;
            }
        }
        else break;
    }
}

static void term_signal(int sig)
{
    running = 0;
    socket_finish();

    /* unused */
    return; (void)sig;
}

static int mod_host_init(jack_client_t* client, int socket_port)
{
#ifdef HAVE_FFTW335
    /* Make fftw thread-safe */
    fftw_make_planner_thread_safe();
    fftwf_make_planner_thread_safe();
#endif
#ifdef HAVE_NE10
    /* Initialize ne10 */
    ne10_init();
#endif

    /* Setup the protocol */
    protocol_add_command(EFFECT_ADD, effects_add_cb);
    protocol_add_command(EFFECT_REMOVE, effects_remove_cb);
    protocol_add_command(EFFECT_PRESET_LOAD, effects_preset_load_cb);
    protocol_add_command(EFFECT_PRESET_SAVE, effects_preset_save_cb);
    protocol_add_command(EFFECT_PRESET_SHOW, effects_preset_show_cb);
    protocol_add_command(EFFECT_CONNECT, effects_connect_cb);
    protocol_add_command(EFFECT_DISCONNECT, effects_disconnect_cb);
    protocol_add_command(EFFECT_BYPASS, effects_bypass_cb);
    protocol_add_command(EFFECT_PARAM_SET, effects_set_param_cb);
    protocol_add_command(EFFECT_PARAM_GET, effects_get_param_cb);
    protocol_add_command(EFFECT_PARAM_MON, effects_monitor_param_cb);
    protocol_add_command(MONITOR_ADDR_SET, monitor_addr_set_cb);
    protocol_add_command(MONITOR_OUTPUT, monitor_output_cb);
    protocol_add_command(MIDI_LEARN, midi_learn_cb);
    protocol_add_command(MIDI_MAP, midi_map_cb);
    protocol_add_command(MIDI_UNMAP, midi_unmap_cb);
    protocol_add_command(MIDI_PROGRAM_LISTEN, midi_program_listen_cb);
    protocol_add_command(CPU_LOAD, cpu_load_cb);
    protocol_add_command(LOAD_COMMANDS, load_cb);
    protocol_add_command(SAVE_COMMANDS, save_cb);
    protocol_add_command(BUNDLE_ADD, bundle_add);
    protocol_add_command(BUNDLE_REMOVE, bundle_remove);
    protocol_add_command(DATA_HANDLED, data_handled);

    /* skip help and quit for internal client */
    if (client == NULL)
    {
        protocol_add_command(HELP, help_cb);
        protocol_add_command(QUIT, quit_cb);
    }

    /* Startup the effects */
    if (effects_init(client))
        return -1;

    /* Setup the socket */
    if (socket_start(socket_port, SOCKET_MSG_BUFFER_SIZE) < 0)
        return -1;

    socket_set_receive_cb(protocol_parse);

    return 0;
}

static void* intclient_socket_run(void* ptr)
{
    while (running)
        socket_run(0);

    /* unused */
    return NULL;
    (void)ptr;
}


/*
************************************************************************************************************************
*           MAIN FUNCTION
************************************************************************************************************************
*/

int main(int argc, char **argv)
{
    /* Command line options */
    static struct option long_options[] = {
        {"nofork", no_argument, 0, 'n'},
        {"verbose", no_argument, 0, 'v'},
        {"socket-port", required_argument, 0, 'p'},
        {"interactive", no_argument, 0, 'i'},
        {"version", no_argument, 0, 'V'},
        {"help", no_argument, 0, 'h'},
        {0, 0, 0, 0}
    };

    int opt, opt_index = 0;

    /* parse command line options */
    int nofork = 0, verbose = 0, socket_port = SOCKET_DEFAULT_PORT, interactive = 0;
    while ((opt = getopt_long(argc, argv, "nvp:iVh", long_options, &opt_index)) != -1)
    {
        switch (opt)
        {
            case 'n':
                nofork = 1;
                break;

            case 'v':
                verbose = 1;
                nofork = 1;
                break;

            case 'p':
                socket_port = atoi(optarg);
                break;

            case 'i':
                interactive = 1;
                nofork = 1;
                break;

            case 'V':
                printf(
                    "%s version: %s\n"
                    "source code: https://github.com/moddevices/mod-host\n",
                argv[0], version);

                exit(EXIT_SUCCESS);
                break;

            case 'h':
                printf(
                    "Usage: %s [-vih] [-p <port>]\n"
                    "  -v, --verbose                  verbose messages\n"
                    "  -p, --socket-port=<port>       socket port definition\n"
                    "  -i, --interactive              interactive mode\n"
                    "  -V, --version                  print program version and exit\n"
                    "  -h, --help                     print this help and exit\n",
                argv[0]);

                exit(EXIT_SUCCESS);
        }
    }

    if (! nofork)
    {
        int pid;
        pid = fork();
        if (pid != 0)
        {
            printf("Forking... child PID: %d\n", pid);

            FILE *fd;
            fd = fopen(PID_FILE, "w");
            if (fd == NULL)
            {
                fprintf(stderr, "can't open PID File\n");
            }
            else
            {
                fprintf(fd, "%d\n", pid);
                fclose(fd);
            }
            exit(EXIT_SUCCESS);
        }
    }

    if (mod_host_init(NULL, socket_port) != 0)
    {
        exit(EXIT_FAILURE);
        return 1;
    }

    /* Interactice mode */
    if (interactive)
    {
        interactive_mode();
    }
    else
    {
        struct sigaction sig;
        memset(&sig, 0, sizeof(sig));

        sig.sa_handler = term_signal;
        sig.sa_flags   = SA_RESTART;
        sigemptyset(&sig.sa_mask);
        sigaction(SIGTERM, &sig, NULL);
        sigaction(SIGINT, &sig, NULL);
    }

    /* Verbose */
    protocol_verbose(verbose);

    /* Report ready */
    printf("mod-host ready!\n");
    fflush(stdout);

    running = 1;
    while (running) socket_run(interactive);

    socket_finish();
    effects_finish(1);
    protocol_remove_commands();

    return 0;
}

__attribute__ ((visibility("default")))
int jack_initialize(jack_client_t* client, const char* load_init);

int jack_initialize(jack_client_t* client, const char* load_init)
{
    int socket_port = SOCKET_DEFAULT_PORT;

    if (load_init != NULL && load_init[0] != '\0')
        socket_port = atoi(load_init);

    if (mod_host_init(client, socket_port) != 0)
        return 1;

    running = 1;
    pthread_create(&intclient_socket_thread, NULL, intclient_socket_run, NULL);

    return 0;
}

__attribute__ ((visibility("default")))
void jack_finish(void);

void jack_finish(void)
{
    running = 0;
    socket_finish();
    pthread_join(intclient_socket_thread, NULL);
    effects_finish(0);
    protocol_remove_commands();
}
